tests/email/test_tokens.py

```diff
@@ -41,177 +41,205 @@ def password_reset_token():
-        data = decode_token(token, 'Secret', 'PassSalt')
+        # Personally identifiable information should not be readable without secret key
+        assert b'test@example.com' not in base64.urlsafe_b64decode(token)
 
-    assert data == ({'email': 'test@example.com', 'user': 123}, datetime(2016, 1, 1, 12, 0, 0))
+        # a fernet string contains the timestamp at which it was encrypted
+        assert _parse_fernet_timestamp(token) == datetime(2016, 1, 1, 12)
 
+    def test_decrypt_token_ok_for_known_good_token(self):
+        # encrypted on 2016-01-01T12:00:00.30Z with secret 'Secret' and namespace 'PassSalt'
+        token = 'gAAAAABWhmpA1ecLpuzdKiIcJ_drdA1Vf4ip07TH3UqZ_fA-pD9yYlGMqSTi-Mpbd58Z-wlZfa5sXGE6FVHPilTpsZWEiMDRLdCBlccvBPbY9IOO5F3uabjkrk87mrlPxaSbMAza5Nku'  # noqa
 
-def test_decode_password_reset_token_ok_for_good_token(email_app, data_api_client, password_reset_token):
-    with email_app.app_context():
-        token = generate_token(password_reset_token, "Key", 'PassSalt')
-        assert decode_password_reset_token(token, data_api_client) == password_reset_token
-    data_api_client.get_user.assert_called_once_with(123)
+        with freeze_time('2016-01-01T12:00:00.30Z'):
+            data = decode_token(token, 'Secret', 'PassSalt')
 
+        assert data == ({'email': 'test@example.com', 'user': 123}, datetime(2016, 1, 1, 12, 0, 0))
 
-def test_decode_password_reset_token_does_not_work_if_bad_token(email_app, data_api_client, password_reset_token):
-    token = generate_token(password_reset_token, "Key", 'PassSalt')[1:]
 
-    with email_app.app_context():
-        assert decode_password_reset_token(token, data_api_client) == {'error': 'token_invalid'}
+class TestDecodePasswordReset:
 
+    def test_decode_password_reset_token_ok_for_good_token(self, email_app, data_api_client, password_reset_token):
+        with email_app.app_context():
+            token = generate_token(password_reset_token, "Key", 'PassSalt')
+            assert decode_password_reset_token(token, data_api_client) == password_reset_token
+        data_api_client.get_user.assert_called_once_with(123)
 
-def test_decode_password_reset_token_does_not_work_if_token_expired(email_app, data_api_client, password_reset_token):
-    with freeze_time('2015-01-02 03:04:05'):
-        # Token was generated a year before current time
-        token = generate_token(password_reset_token, "Key", 'PassSalt')
+    def test_decode_password_reset_token_does_not_work_if_bad_token(
+            self, email_app, data_api_client, password_reset_token):
+        token = generate_token(password_reset_token, "Key", 'PassSalt')[1:]
 
-    with freeze_time('2016-01-02 03:04:05'):
         with email_app.app_context():
             assert decode_password_reset_token(token, data_api_client) == {'error': 'token_invalid'}
 
+    @pytest.mark.parametrize(
+        'decode_time, expected_result', [
+            ('2016-01-02 03:04:05', 'ok'),
```

Comment:
Would be good to test `('2016-01-02 03:04:04', 'error'),` as well.

Reply:
Interesting! We don't fail when the token is generated in the future (i.e. after we decode the token... that we are about to generate). I... don't know how this would happen in practice?

Reply:
I've added in an extra line to make it explicit that this shouldn't happen.

Reply:
I think we're OK - the Fernet timestamp is UTC (https://github.com/fernet/spec/blob/master/Spec.md), and we're using UTC everywhere too.